### ISSUE TITLE
Add sanity check for libraries with missing symbols

### ISSUE THREAD

**Issue by adamjstewart** [spack/spack#1304]
I've come across two packages now that have built without any trouble, but have later caused problems for users trying to link to them. The first was py-meep, which complained about missing symbols when imported (see https://github.com/LLNL/spack/pull/1156#issuecomment-233030464). The second was Armadillo (see https://github.com/LLNL/spack/issues/1299#issue-166378265).

These problems can be hard to debug, but fairly easy to check for. In the case of Armadillo, it looks like this:

``` bash
$ ldd -r libarmadillo.so
    linux-vdso.so.1 =>  (0x00007ffdd8b61000)
    libopenblas.so.0 => /blues/gpfs/home/software/spack/opt/spack/linux-centos6-x86_64/gcc-6.1.0/openblas-0.2.18-avx2mvh76o62v5a2dr5u6zz2ewer5p7w/lib/libopenblas.so.0 (0x00002b8a88ba1000)
    libstdc++.so.6 => /blues/gpfs/home/software/spack/opt/spack/linux-x86_64/gcc-5.3.0/gcc-6.1.0-q2zosj3igepi3pjnqt74bwazmptr5gpj/lib64/libstdc++.so.6 (0x00002b8a89b7f000)
    libm.so.6 => /lib64/libm.so.6 (0x00002b8a89f16000)
    libgcc_s.so.1 => /blues/gpfs/home/software/spack/opt/spack/linux-x86_64/gcc-5.3.0/gcc-6.1.0-q2zosj3igepi3pjnqt74bwazmptr5gpj/lib64/libgcc_s.so.1 (0x00002b8a8a19a000)
    libc.so.6 => /lib64/libc.so.6 (0x00002b8a8a3b1000)
    /lib64/ld-linux-x86-64.so.2 (0x00002b8a886f1000)
    libpthread.so.0 => /lib64/libpthread.so.0 (0x00002b8a8a745000)
    libgfortran.so.3 => /blues/gpfs/home/software/spack/opt/spack/linux-x86_64/gcc-5.3.0/gcc-6.1.0-q2zosj3igepi3pjnqt74bwazmptr5gpj/lib64/libgfortran.so.3 (0x00002b8a8a963000)
    libquadmath.so.0 => /blues/gpfs/home/software/spack/opt/spack/linux-x86_64/gcc-5.3.0/gcc-6.1.0-q2zosj3igepi3pjnqt74bwazmptr5gpj/lib64/../lib64/libquadmath.so.0 (0x00002b8a8ac8a000)
undefined symbol: zvout_    (./libarmadillo.so)
undefined symbol: dvout_    (./libarmadillo.so)
undefined symbol: cvout_    (./libarmadillo.so)
undefined symbol: dmout_    (./libarmadillo.so)
undefined symbol: smout_    (./libarmadillo.so)
undefined symbol: cmout_    (./libarmadillo.so)
undefined symbol: svout_    (./libarmadillo.so)
undefined symbol: zmout_    (./libarmadillo.so)
undefined symbol: ivout_    (./libarmadillo.so)
```

I propose the following sanity check. Run this command on all installed dynamic libraries:

``` bash
ldd -r <libname.so> | grep "undefined symbol"
```

@citibeth You've been the biggest opponent of sanity checks. Can you foresee any problems with this?

**Comment by alalazo** [comment]
I agree with this. Anyhow, if we are making this a feature that gets shipped as a default we should be sure to support potentially different strategy for different `platforms` before merging something that will work only for usual linux machines...

**Comment by tgamblin** [comment]
Missing symbols and also libs outside the Spack directory would be useful.  I pushed (a long time ago) a prototype for the latter in [features/libs-command](https://github.com/LLNL/spack/tree/features/libs-command).

**Comment by citibeth** [comment]
As others have pointed out, this proposal is Linux-specific.

This sanity check would certainly be more useful than the ones I have experience with in the past.  EasyBuild would just check for the existence of a `.so` file.  This initially broke everything on Mac, where they were named `.dylib`.

The purist in me wonders if this stuff should really be part of Spack at all.  Spack's job is to set up builds and tell them where to install.  It shouldn't have to get involved in the details of the artifacts installed; why are packages willing to build libraries with missing symbols?  Ideally, the build procedures in the packages would sanity-check themselves.  We sometimes use Spack to compensate for deficiencies in a build (see gtest/googletest package, for example, which has no install procedure).  But it never quite feels right to me.

However, the practical side of me sees the benefits here.  If we can rename log files to `.txt`, we can sanity-check shared libraries as well.

Just as important as the content of the sanity check, is how it is managed.  If the check fails, the user should have a way to proceed, *without rebuilding*.  That's what was worst about EasyBuild sanity checks... your package would spend 15 minutes building successfully, a broken sanity check would fail, and then you'd have to fix the sanity check and build for another 15 minutes.  If your fix had an error... well, wait another 15 minutes to get it right.

**Comment by adamjstewart** [comment]
@citibeth Now I'm starting to see the point you were trying to make when you said you wanted to be able to install with `--run-tests`, but have the installation complete even if the tests fail. 

What if you ran `spack install --run-tests`. If the tests crash, the installation would crash and you could view the test log files to determine whether or not you really care. Then, if you decide they aren't important, you could re-run `spack install` without `--run-tests`, and Spack would pick up where it left off, installing the package for you. That way you don't have to wait for another 15 minute rebuild. @alalazo How feasible do you think this would be?

**Comment by citibeth** [comment]
Sounds fine with me.  As long as Spack doesn't throw away the 15 minutes
spent building.

**Comment by boegel** [comment]
@citibeth:
> Just as important as the content of the sanity check, is how it is managed. If the check fails, the > user should have a way to proceed, without rebuilding. That's what was worst about EasyBuild sanity checks... your package would spend 15 minutes building successfully, a broken sanity check would fail, and then you'd have to fix the sanity check and build for another 15 minutes. If your fix had an error... well, wait another 15 minutes to get it right.

This is not correct, at least not since we introduced `--module-only` (April'15), see also http://easybuild.readthedocs.io/en/latest/Partial_installations.html#only-re-generating-additional-module-files-using-module-only

**Comment by adamjstewart** [comment]
I'm starting to think that this test isn't such a good idea. I've come across some packages (flex?) that _should_ have missing symbols. It is up to you as a developer using this package to define those symbols.

**Comment by junghans** [comment]
Sorry for being late to the party!

If one injects `-Wl,-z,defs` (for gcc / GNU binutils) into the `CXXFLAGS` linking will fails for libraries with undefined symbols. I guess we could do that automatically in some sort of maintainer mode.

**Comment by alalazo** [comment]
Fixed by #47365
